// express-openapi-generator/lib/types/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentFieldNames = exports.CompositeSchemaTypes = void 0;
var CompositeSchemaTypes;
(function (CompositeSchemaTypes) {
    CompositeSchemaTypes["allOf"] = "allOf";
    CompositeSchemaTypes["oneOf"] = "oneOf";
    CompositeSchemaTypes["anyOf"] = "anyOf";
})(CompositeSchemaTypes = exports.CompositeSchemaTypes || (exports.CompositeSchemaTypes = {}));
var ComponentFieldNames;
(function (ComponentFieldNames) {
    ComponentFieldNames["schemas"] = "schemas";
    ComponentFieldNames["responses"] = "responses";
    ComponentFieldNames["parameters"] = "parameters";
    ComponentFieldNames["examples"] = "examples";
    ComponentFieldNames["requestBodies"] = "requestBodies";
    ComponentFieldNames["headers"] = "headers";
    ComponentFieldNames["securitySchemes"] = "securitySchemes";
    ComponentFieldNames["links"] = "links";
    ComponentFieldNames["callbacks"] = "callbacks";
})(ComponentFieldNames = exports.ComponentFieldNames || (exports.ComponentFieldNames = {}));
